signedauth/signatureauth.go

```diff
@@ -0,0 +1,89 @@
+package signedauth
+
+import (
+	"crypto/hmac"
+	"encoding/hex"
+	"errors"
+	"github.com/gin-gonic/gin"
+	"hash"
+	"net/http"
+	"strings"
+)
+
+// SignedAuthManager defines the functions needed to fulfill an auth key managing role.
+type SignedAuthManager interface {
+	AuthHeaderPrefix() string                         // The beginning of the string from the HTTP_AUTHORIZATION header. The exact header must be followed by a space.
+	SecretKey(string, *http.Request) (string, *Error) // The secret key for the provided access key and request. Header verification should happen here, and an error returned to fail.
+	DataToSign(*http.Request) (string, *Error)        // The data which must be signed and verified, or an error to return.
+	AuthHeaderRequired() bool                         // Whether or not a request without any header should be accepted (c.Next) or forbidden (c.Fail with status 403).
+	HashFunction() func() hash.Hash                   // Returns the hash function to use, e.g. sha1.New (imported from "crypto/sha1"), or sha512.New384 for SHA-384.
+	ContextKey() string                               // The key in the context where will be set the appropriate value if the request was correctly signed.
+	ContextValue(string) interface{}                  // The value which will be stored in the context if authentication is successful, from the access key.
+}
+
+// Error defines the authentication failure with a status. The error string will *not* be returned by Gin.
+type Error struct {
+	Status int   // The status for this failure.
+	Err    error // The error associated to this failure.
+}
+
+// SignatureAuth is the middleware function. It must be called with a struct which implements the SignedAuthManager interface.
+func SignatureAuth(mgr SignedAuthManager) gin.HandlerFunc {
+
+	return func(c *gin.Context) {
+		accesskey, signature, err := extractAuthInfo(mgr, c.Request.Header.Get("Authorization"))
+		if err != nil {
+			// Credentials doesn't match, we return 401 Unauthorized and abort request.
+			c.Fail(err.Status, err.Err)
+		} else if accesskey == "" && signature == "" && !mgr.AuthHeaderRequired() {
+			c.Next()
+		} else {
+			// Authorization header has the correct format.
+			secret, keyerr := mgr.SecretKey(accesskey, c.Request)
+			if keyerr != nil {
+				c.Fail(keyerr.Status, keyerr.Err)
+			} else {
+				data, dataerr := mgr.DataToSign(c.Request)
+				if dataerr != nil {
+					c.Fail(dataerr.Status, dataerr.Err)
+				} else if !isSignatureValid(mgr.HashFunction(), secret, data, signature) {
+					// Accesskey is valid but signature is not.
+					c.Fail(http.StatusUnauthorized, errors.New("Wrong access key or signature."))
+				} else {
+					// Accesskey and signature are valid.
+					c.Set(mgr.ContextKey(), mgr.ContextValue(accesskey))
+					c.Next()
```

Comment:
Thanks for your comments. I'm just starting out with Go, so I'll fix these.
As for the Next instruction, I took that code from the auth middleware I
believe.

On Fri, May 22, 2015, 19:40 Manu Mtz.-Almeida <notifications@github.com>
wrote:

> In signedauth/signatureauth.go
> <https://github.com/gin-gonic/contrib/pull/44#discussion_r30920507>:
>
> > +		} else {
> > +			// Authorization header has the correct format.
> > +			secret, keyerr := mgr.SecretKey(accesskey, c.Request)
> > +			if keyerr != nil {
> > +				c.Fail(keyerr.Status, keyerr.Err)
> > +			} else {
> > +				data, dataerr := mgr.DataToSign(c.Request)
> > +				if dataerr != nil {
> > +					c.Fail(dataerr.Status, dataerr.Err)
> > +				} else if !isSignatureValid(mgr.HashFunction(), secret, data, signature) {
> > +					// Accesskey is valid but signature is not.
> > +					c.Fail(http.StatusUnauthorized, errors.New("Wrong access key or signature."))
> > +				} else {
> > +					// Accesskey and signature are valid.
> > +					c.Set(mgr.ContextKey(), mgr.ContextValue(accesskey))
> > +					c.Next()
>
> we do not have to call c.Next() in order to continue the chain
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gin-gonic/contrib/pull/44/files#r30920507>.
>